### PR TITLE
add readiness probe to gloo and gateway deployments

### PR DESCRIPTION
**Pull request by ilackarms**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1322

### CODE REVIEW

**Comment by ilackarms** [comment]
currently some of the e2e tests fail because validation is in an un-ready state when config resources start getting created.

this pr adds a readiness probe to both Gloo and Gateway (simple [tcp readiness probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-tcp-liveness-probe)) which ensure their http servers have started listening.

this, in combination with the `glooctl check` that we run in the `BeforeEach` ensure that both the webhook and the proxy validation servers are online before beginning our tests

unfortunately this adds some cost (about 10-15s per kube2e gateway test) on to the test runs. we should consider running all e2e tests against a single installation of gloo rather than reinstalling in a BeforeEach (any problem with using a `BeforeSuite`?)

File: changelog/v0.20.3/fix-validation-test-flake.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+changelog:
+  - type: FIX
+    description: Gateway waits for Gloo validation server to become available during setup
+    issueLink: https://github.com/solo-io/gloo/issues/
```
Comment (marcogschmidt) [inline changelog/v0.20.3/fix-validation-test-flake.yaml]: URL seems incomplete

File: install/helm/gloo/templates/1-gloo-deployment.yaml [hunk 2]
```diff
@@ -69,6 +69,9 @@ spec:
           - name: START_STATS_SERVER
             value: "true"
         {{- end}}
+        readinessProbe:
+          tcpSocket:
+            port: {{ .Values.gloo.deployment.xdsPort }}
```
Comment (marcogschmidt) [inline install/helm/gloo/templates/1-gloo-deployment.yaml]: Have you considered adding `periodSeconds` (probe frequency)? The default is 10 seconds, maybe by bringing it down to a lower value and increasing the `failureThreshold` we can shave off some some seconds from each test run.

File: projects/gateway/pkg/syncer/setup_syncer.go [hunk 3]
```diff
@@ -213,24 +213,29 @@ func RunGateway(opts Opts) error {
...
+	)
 	var ignoreProxyValidationFailure bool
 	var allowMissingLinks bool
 	if opts.Validation != nil {
-		contextutils.LoggerFrom(ctx).Infow("starting proxy validation client... this may take a moment",
-			zap.String("validation_server", opts.Validation.ProxyValidationServerAddress))
-		cc, err := grpc.DialContext(ctx, opts.Validation.ProxyValidationServerAddress, grpc.WithInsecure(), grpc.WithBlock())
-		if err != nil {
-			return errors.Wrapf(err, "failed to initialize grpc connection to validation server.")
+		validationClientConstructor = func() (client validation.ProxyValidationServiceClient, e error) {
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/gateway/pkg/syncer/setup_syncer.go]: i don't think it should actually. that means each call will get its own grpc client.

the constructor here wraps the context from the original `Setup`

**Comment by ilackarms** [comment]
/kick
passed ci once, testing for flakes

**Comment by ilackarms** [comment]
/kick

test flake? 

```
Step #11 - "test": • Failure [0.777 seconds]
Step #11 - "test": Knative RenderKnativeManifests [It] renders manifests for each knative component 
Step #11 - "test": /workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/install/knative_test.go:27
Step #11 - "test": 
Step #11 - "test":   Unexpected error:
Step #11 - "test":       <*xerrors.noWrapError | 0xc000ec2a40>: {
Step #11 - "test":           msg: "returned non-200 status code: 503 503 Slow Down",
Step #11 - "test":           err: nil,
Step #11 - "test":           frame: {
Step #11 - "test":               frames: [0x176f21f, 0x27d8e8d, 0x27d8e58],
Step #11 - "test":           },
Step #11 - "test":       }
Step #11 - "test":       returned non-200 status code: 503 503 Slow Down
Step #11 - "test":   occurred
Step #11 - "test": 
Step #11 - "test":   /workspace/gopath/src/github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:48
Step #11 - "test": 
Step #11 - "test":   Full Stack Trace
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/vendor/github.com/solo-io/go-utils/testutils/fail_handler.go:51 +0x17e
Step #11 - "test":   github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion.(*Assertion).match(0xc000ec2a80, 0x34359e0, 0x5c07ed0, 0x0, 0x0, 0x0, 0x0, 0x2bfad80)
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:75 +0x2e2
Step #11 - "test":   github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion.(*Assertion).NotTo(0xc000ec2a80, 0x34359e0, 0x5c07ed0, 0x0, 0x0, 0x0, 0x2ef4e12)
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:48 +0x11c
Step #11 - "test":   github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/install.glob..func1.1.1()
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/install/knative_test.go:29 +0x278
Step #11 - "test":   github.com/solo-io/gloo/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0007c65a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/install/install_suite_test.go:18 +0x256
Step #11 - "test":   testing.tRunner(0xc00005f000, 0x2fce4e0)
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:865 +0x164
Step #11 - "test":   created by testing.(*T).Run
```

File: projects/gateway/pkg/validation/robust_client.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+// if a connection error occurs during an api call, the robustValidationClient
+// attempts to reestablish the connection & retry the call before returning the error
+type robustValidationClient struct {
+	lock                      sync.RWMutex
+	validationClient          validation.ProxyValidationServiceClient
+	constructValidationClient func() (validation.ProxyValidationServiceClient, error)
+}
+
+// the constructor returned here is not threadsafe; call from a lock
+func RetryOnUnavailableClientConstructor(ctx context.Context, serverAddress string) ClientConstructor {
```
Comment (marcogschmidt) [inline projects/gateway/pkg/validation/robust_client.go]: Love that you moved this to a dedicated struct and documented it 👏

File: projects/gateway/pkg/validation/robust_client.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		cc, err := grpc.DialContext(clientCtx, serverAddress, grpc.WithInsecure(), grpc.WithBlock())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to initialize grpc connection to validation server.")
+		}
+
+		return validation.NewProxyValidationServiceClient(cc), nil
+	}
+}
+
+func NewRobustValidationClient(constructValidationClient func() (validation.ProxyValidationServiceClient, error)) (*robustValidationClient, error) {
```
Comment (marcogschmidt) [inline projects/gateway/pkg/validation/robust_client.go]: Nit:
```suggestion
func NewRobustValidationClient(constructValidationClient ClientConstructor) (*robustValidationClient, error) {
```

File: projects/gateway/pkg/validation/robust_client.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		}
+	})); err != nil {
+		return nil, err
+	}
+	return proxyReport, nil
+}
+
+func isUnavailableErr(err error) bool {
+	if err == nil {
+		return true
```
Comment (marcogschmidt) [inline projects/gateway/pkg/validation/robust_client.go]: This never occurs (checked the retry code and `err` is always non-nil), but it should return `false`. We can just delete this check and the `switch` below will cover us.

File: projects/gateway/pkg/validation/robust_client.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	var reinstantiateClientErr error
+	if err := retry.Do(func() error {
+		c.lock.RLock()
+		defer c.lock.RUnlock()
+		validationClient = c.validationClient
+		var err error
+		proxyReport, err = validationClient.ValidateProxy(ctx, proxy, opts...)
+		return err
+	}, retry.RetryIf(func(e error) bool {
+		if reinstantiateClientErr != nil {
```
Comment (marcogschmidt) [inline projects/gateway/pkg/validation/robust_client.go]: I checked the retry library and, contrary to what we thought, `retryIf` gets called [**after**](https://github.com/avast/retry-go/blob/101acd3290b786bf437607d874596e3996484d8d/retry.go#L103) `onRetry`. This means that if we get a `reinstantiateClientErr`, we will not retry! I think a fix would be to remove this check.

Comment (marcogschmidt) [reply projects/gateway/pkg/validation/robust_client.go]: Nvm, we wrap this whole call in a retry. Also you rightly pointed out that retrying when `reinstantiateClientErr != nil` would throw a nil pointer deref.

**Pull request merged**